Content.Shared/Pinpointer/SharedNavMapSystem.cs

```diff
@@ -31,19 +56,235 @@ public static Vector2i GetTile(int flag)
         return new Vector2i(x, y);
     }
 
-    [Serializable, NetSerializable]
-    protected sealed class NavMapComponentState : ComponentState
+    public NavMapChunk SetAllEdgesForChunkTile(NavMapChunk chunk, Vector2i tile)
+    {
+        var relative = SharedMapSystem.GetChunkRelative(tile, ChunkSize);
+        var flag = (ushort) GetFlag(relative);
+
+        foreach (var (direction, _) in chunk.TileData)
+            chunk.TileData[direction] |= flag;
+
+        return chunk;
+    }
+
+    public NavMapChunk UnsetAllEdgesForChunkTile(NavMapChunk chunk, Vector2i tile)
+    {
+        var relative = SharedMapSystem.GetChunkRelative(tile, ChunkSize);
+        var flag = (ushort) GetFlag(relative);
+        var invFlag = (ushort) ~flag;
+
+        foreach (var (direction, _) in chunk.TileData)
+            chunk.TileData[direction] &= invFlag;
+
+        return chunk;
+    }
+
+    public ushort GetCombinedEdgesForChunk(Dictionary<AtmosDirection, ushort> tile)
+    {
+        ushort combined = 0;
+
+        foreach (var (_, value) in tile)
+            combined |= value;
+
+        return combined;
+    }
+
+    public bool AllTileEdgesAreOccupied(Dictionary<AtmosDirection, ushort> tileData, Vector2i tile)
+    {
+        var flag = (ushort) GetFlag(tile);
+
+        foreach (var (direction, _) in tileData)
+        {
+            if ((tileData[direction] & flag) == 0)
+                return false;
+        }
+
+        return true;
+    }
+
+    public NavMapChunkType GetAssociatedEntityChunkType(EntityUid uid)
+    {
+        var category = NavMapChunkType.Invalid;
+
+        if (HasComp<NavMapDoorComponent>(uid))
+            category = NavMapChunkType.Airlock;
+
+        else if (_tagSystem.HasAnyTag(uid, _wallTags))
+            category = NavMapChunkType.Wall;
+
+        return category;
+    }
+
+    protected bool TryCreateNavMapBeaconData(EntityUid uid, NavMapBeaconComponent component, TransformComponent xform, [NotNullWhen(true)] out NavMapBeacon? beaconData)
     {
-        public Dictionary<Vector2i, int> TileData = new();
+        beaconData = null;
 
-        public List<NavMapBeacon> Beacons = new();
+        if (!component.Enabled || xform.GridUid == null || !xform.Anchored)
+            return false;
 
-        public List<NavMapAirlock> Airlocks = new();
+        string? name = component.Text;
+
+        if (string.IsNullOrEmpty(name))
+            name = MetaData(uid).EntityName;
+
+        beaconData = new NavMapBeacon(GetNetEntity(uid), component.Color, name, xform.LocalPosition)
+        {
+            LastUpdate = _gameTiming.CurTick
```

Comment:
You should be able to use the metadatacomponent across both the entname and getnetentity.